src/sourmash/index.py

```diff
@@ -219,7 +220,66 @@ def select(self, **kwargs):
             if select_signature(ss, **kwargs):
                 siglist.append(ss)
 
-        return LinearIndex(siglist, self.filename)
+        return LinearIndex(siglist, self.location)
+
+
+class ZipFileLinearIndex(Index):
+    """\
+    A read-only collection of signatures in a zip file.
+
+    Does not support `insert` or `save`.
+    """
+    is_database = True
+
+    def __init__(self, zf, selection_dict=None,
+                 traverse_yield_all=False):
+        self.zf = zf
+        self.selection_dict = selection_dict
+        self.traverse_yield_all = traverse_yield_all
+
+    def __len__(self):
+        return len(list(self.signatures()))
+
+    @property
+    def location(self):
+        return self.zf.filename
+
+    def insert(self, signature):
+        raise NotImplementedError
+
+    def save(self, path):
+        raise NotImplementedError
+
+    @classmethod
+    def load(cls, location, traverse_yield_all=False):
+        "Class method to load a zipfile."
+        zf = zipfile.ZipFile(location, 'r')
+        return cls(zf, traverse_yield_all=traverse_yield_all)
+
+    def signatures(self):
+        "Load all signatures in the zip file."
+        from .signature import load_signatures
+        for zipinfo in self.zf.infolist():
+            # should we load this file? if it ends in .sig OR we are forcing:
+            if zipinfo.filename.endswith('.sig') or \
+               zipinfo.filename.endswith('.sig.gz') or \
+               self.traverse_yield_all:
+                fp = self.zf.open(zipinfo)
+
+                # now load all the signatures and select on ksize/moltype:
+                selection_dict = self.selection_dict
+                for ss in load_signatures(fp):
```

Comment:
Why `load_signatures` rather than `load_file_as_signatures`, which would enable loading index files that are nested in the zipfile? Do we want to avoid potentially complicated nesting?

Reply:
good question! `load_file_as_signatures` takes a filename, while `load_signatures` can take a filename, a string, or a file handle - which is what it's doing here.

Reply:
ok! Guess I'm a little confused on deprecation of `load_signatures`, then --what is the plan for this functionality in the future? Or are we just deprecating from the public api?

Reply:
:) good question. Yes, deprecating from the public API, and also --

* maybe changing `load_file_as_signatures` to support file handles https://github.com/dib-lab/sourmash/issues/1312
* breaking up `load_signatures` to be less complicated, implementation-wise https://github.com/dib-lab/sourmash/issues/1060

Reply:
thx!